src/clout/core.cljc

```diff
@@ -0,0 +1,162 @@
+                         (path-info request))]
+         (if (.test re path-info)
+           (assoc-keys-with-groups (re-groups* re path-info) keys)))))
+  #?@(:clj [Object
+            (toString [_] source)]))
+
+#?(:clj
+   (def ^:private route-parser
+     (insta/parser
+       "route    = (scheme / part) part*
+        scheme   = #'(https?:)?//'
+        <part>   = literal | escaped | wildcard | param
+        literal  = #'(:[^\\p{L}_*{}\\\\]|[^:*{}\\\\])+'
+        escaped  = #'\\\\.'
+        wildcard = '*'
+        param    = key pattern?
+        key      = <':'> #'([\\p{L}_][\\p{L}_0-9-]*)'
+        pattern  = '{' (#'(?:[^{}\\\\]|\\\\.)+' | pattern)* '}'"
+       :no-slurp true))
+   :cljs
+   (def ^:private route-parser
+     ;; it is too complicated to implement \p{L} in cljs
+     ;; A-Za-z is pretty enough for keywords
+     (insta/parser
+       "route    = (scheme / part) part*
+        scheme   = #'(https?:)?//'
+        <part>   = literal | escaped | wildcard | param
+        literal  = #'(:[^A-Za-z_*{}\\\\]|[^:*{}\\\\])+'
+        escaped  = #'\\\\.'
+        wildcard = '*'
+        param    = key pattern?
+        key      = <':'> #'([A-Za-z_][A-Za-z_0-9-]*)'
+        pattern  = '{' (#'(?:[^{}\\\\]|\\\\.)+' | pattern)* '}'"
+       :no-slurp true)))
+
+(defn- parse [parser text]
+  (let [result (insta/parse parser text)]
+    (if (insta/failure? result)
+      (throw (ex-info "Parse error in route string" {:failure result}))
+      result)))
+
+(defn- find-route-key [form]
+  (case (first form)
+    :wildcard :*
+    :param (-> form second second keyword)))
+
+(defn- route-keys [parse-tree]
+  (->> (rest parse-tree)
+       (filter (comp #{:param :wildcard} first))
+       (map find-route-key)))
+
+(defn- trim-pattern [pattern]
+  (some-> pattern (subs 1 (dec (count pattern)))))
+
+(defn- param-regex [regexs key & [pattern]]
+  (str "(" (or (trim-pattern pattern) (regexs key) "[^/,;?]+") ")"))
+
+(defn- route-regex [parse-tree regexs]
+  (insta/transform
+    {:route    (comp re-pattern #?(:cljs #(str "^" % "$")) str)
+     :scheme   #(if (= % "//") "https?://" %)
+     :literal  re-escape
+     :escaped  #(re-escape (subs % 1))
+     :wildcard (constantly "(.*?)")
+     :param    (partial param-regex regexs)
+     :key      keyword
+     :pattern  str}
+    parse-tree))
+
+(defn- absolute-url? [path]
+  (boolean (re-matches #"(https?:)?//.*" path)))
+
+(defn route-compile
+  "Compile a route string for more efficient route matching."
+  ([path]
+   (route-compile path {}))
+  ([path regexs]
+   (let [ast (parse route-parser path)
+         ks (route-keys ast)]
+     (assert (set/subset? (set (keys regexs)) (set ks))
+             "unused keys in regular expression map")
+     (CompiledRoute.
+       path
+       (route-regex ast regexs)
+       (vec ks)
+       (absolute-url? path)))))
+
+(extend-type #?(:clj String :cljs string)
+  Route
+  (route-matches [route request]
+    (route-matches (route-compile route) request)))
```

Comment:
Files should end with a newline.

Reply:
I fixed this

Reply:
Just one newline is necessary, not two. Can you ensure that the file ends with *one* newline.

Reply:
Yes, sorry